lib/core/src/Cardano/Pool/Performance.hs

```diff
@@ -0,0 +1,198 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE OverloadedLabels #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- |
+-- Copyright: © 2018-2020 IOHK
+-- License: Apache-2.0
+--
+-- This module defines necessary bits to compute an averaged apparent
+-- performance for stake pools. The definition is inspired from the "Design
+-- Specification for Delegation and Incentives in Cardano" by Kant et Al, yet
+-- deviates slightly from it because of different real constraints:
+--
+--     a) The wallet software doesn't have a reliable access to various epoch
+--     details such as the stake distribution for past epochs.
+--
+--     b) The wallet software makes therefore a significant use of the current
+--     epoch's data which represents unfinished snapshots of epochs which, because
+--     of their non-deterministic nature, may favor leaders elected early in the
+--     process more than those not yet elected.
+--
+-- To cope with these two issues, we try to harmonize past known epochs with the
+-- ongoing one by averaging over a whole range of epochs, simultaneously. This
+-- means computing a ratio of the total blocks produced by a leader on the total
+-- number of blocks we could reasonably expect a leader to have produced, across
+-- many epochs at once (i.e., doing a ratio of sums, instead of a sum of
+-- ratios).
+--
+-- This should be less punitive for leaders that have not yet been elected as
+-- part of the ongoing epoch while having still performed reasonably okay in the
+-- past.
+module Cardano.Pool.Performance
+    ( EpochStats (..)
+    , apparentPerformance
+    , readPoolsPerformances
+
+    -- * Helpers
+    , count
+    ) where
+
+
+import Prelude
+
+import Cardano.Pool.DB
+    ( DBLayer (..) )
+import Cardano.Wallet.Primitive.Types
+    ( ActiveSlotCoefficient (..)
+    , EpochLength (..)
+    , EpochNo (..)
+    , PoolId
+    , SlotId (..)
+    , SlotNo (..)
+    )
+import Control.Monad
+    ( forM )
+import Data.Generics.Internal.VL.Lens
+    ( (^.) )
+import Data.Map.Merge.Strict
+    ( dropMissing, traverseMissing, zipWithMatched )
+import Data.Map.Strict
+    ( Map )
+import Data.Quantity
+    ( Quantity (..) )
+import Data.Word
+    ( Word64 )
+import GHC.Generics
+    ( Generic )
+import Numeric.Natural
+    ( Natural )
+
+import qualified Data.Map.Merge.Strict as Map
+import qualified Data.Map.Strict as Map
+
+data EpochStats = EpochStats
+    { poolProduction :: !Natural
+    , poolStake :: !Natural
+    , totalStake :: !Natural
+    , epochHeight :: !Natural
+    } deriving (Generic, Show)
+
+-- We consider the following variables, indexed by a given epoch
+--
+-- - n(e) = number of blocks produced in an epoch e
+-- - N(e) = number of slots in e
```

Comment:
One that jumped to me when looking at the spec today was this:
<img width="781" alt="Skärmavbild 2020-01-17 kl  18 55 38" src="https://user-images.githubusercontent.com/304423/72634627-34de1c00-395b-11ea-8fa7-dfeb52de70c3.png">

Haven't considered the pros/cons, but seems it it recommends a different way than `N` and `activeSlotCoeff`.

Reply:
I am not sure to get your point here?

Reply:
Ok, how about this highlight:

> of **blocks** that were added to the chain during the epoch.

We're dealing with slots here. The spec says blocks.

Reply:
Right! We could change the ` (slotsInEpoch ep)` with the actual sum of blocks produced within the given epoch. And actually get rid of the active slot coefficient :thinking:

Reply:
Fixed!

Reply:
Done :+1: